cpp/ql/lib/semmle/code/cpp/models/interfaces/NonThrowing.qll

```diff
@@ -0,0 +1,13 @@
+/**
+ * Provides an abstract class for modeling functions that never throws.
+ *
+ * See also `ThrowingFunction` for modeling functions that do throw.
```

Comment:
> This is interesting. It looks like ThrowingFunction is for functions that directly throw an exception, i.e. are used like throw itself.

Not necessarily true. When you implement ThrowingFunction you also override a predicate that says whether the throw is unconditional or not.

Now, in practice, we've only implemented it for a single function which happens to throw unconditionally (as this was the motivation for that PR)

Reply:
Oh, I put the note there because I thought that `ThrowingFunction` is the opposite of `NonThrowingFunction`. The description in the documentation for `ThrowingFunction` makes it sound that way to me. But, I can see that the `WindowsDriverFunction` subclass indeed look like `throw`-like constructs. In that case I agree that it shouldn't be mentioned here.

Reply:
I took the note out (and wrote my comment) before I saw @MathiasVP reply. Should I put it back in? _If_ it is the opposite of `NonThrowingFunction` then I think the note is helpful to have.

Reply:
I'm happy that we don't recommend `ThrowingFunction`, _if_ it is supposed to model the exception behaviour of all functions it's extremely incomplete.

Reply:
> [...] if it is supposed to model the exception behaviour of all functions it's extremely incomplete.

It is totally incomplete, yes! We haven't really ever prioritized exception flow (control nor data) since it has basically never been requested other than the one FN we got from Microsoft that motivated the model (see [here](https://github.com/github/codeql/pull/15461) for the PR that did this).